distributed/scheduler.py

```diff
@@ -7982,7 +7982,14 @@ def decide_worker(
     if ts._actor:
         candidates = set(all_workers)
     else:
-        candidates = {wws for dts in deps for wws in dts._who_has}
+        candidates = {
+            wws
+            for dts in deps
+            for wws in dts._who_has
+            # Ignore dependencies that will need to be, or already are, copied to all workers
+            if max(len(dts._who_has), len(dts._dependents))
```

Comment:
I understand why `who_has` is part of this check but not necessarily why dependents is part of this. Having many dependents does not tell us much about where the keys will end up, does it?

Reply:
Yeah, dependents is speculative/heuristic. If we don't include that, the test fails and this PR basically does nothing, except if you've already broadcast a key to every worker before submitting some other tasks.

When the first root task completes and a dep task is getting assigned, the `everywhere` task is still only on one worker. It's actually the process of running the dep tasks that eventually copies `everywhere` everywhere. So without `len(dts._dependents)`, this check won't kick in until we've already made n_workers bad decisions.

The idea is that if a task has lots of dependents, it's quite likely those dependents will run on many different workers. We're trying to guess when this situation will happen before it happens.

There are times ways this heuristic is wrong:
* The many dependents don't have any other dependencies (so running them all on the same worker might actually be a good idea)
  * Though in some cases, they might get identified as root-ish tasks and spread across workers anyway
* The many dependents have worker/resource restrictions

Here is a graph which currently behaves well, but under this PR will allow the `deps` to be assigned to any worker on a 4-worker cluster:
![fantree](https://user-images.githubusercontent.com/3309802/133666019-c7b461f7-ddb7-43e6-8dfd-62e205687e63.png)
Because each root has >4 dependents, the location of the root is ignored when assigning the dependents and they get scattered all over.

I think this heuristic needs more thought. I have another idea I'm going to try out.

Reply:
The other idea (amortizing transfer cost by number of dependencies) didn't work (though it may still be a good idea). The above case works, but then the original test in this PR still fails, because transfer cost is just low regardless, and occupancy is all that really matters.

Interestingly the above case would probably be handled by STA. I still feel like we should tackle STA first, since so many other things will change that it's hard to think about heuristics right now.

Reply:
Okay I got this case working with e1fd58b90bee03003b6f734ce10c780262307d8f, which I don't love, but maybe is okay enough?